### PR TITLE
Allow Millenium Patron API calls to use POST

### PR DESCRIPTION
**Pull request by io7m**
## Description

This adds a configuration option to allow Millenium Patron API requests to be made using POST.

There is an unanswered question here: The original GET requests include some URL encoding. I don't know if we want to have the same encoding in the POST requests.

## Motivation and Context

This is the CM half of the work required to allow requests to use POST. This avoids having passwords logged in URLs.

Affects: https://www.notion.so/lyrasis/Allow-CM-to-verify-password-with-VLC-via-POST-request-e59ab3ca08ed4f279e991cef017fa024

## How Has This Been Tested?

I ran all of the tests locally, and tried POSTing against a local VLC instance.

## Checklist

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

File: api/millenium_patron.py [hunk 1]
```diff
@@ -328,6 +362,13 @@ def request(self, url, *args, **kwargs):
         self._update_request_kwargs(kwargs)
         return HTTP.request_with_timeout("GET", url, *args, **kwargs)
 
+    def request_post(self, url, *args, **kwargs):
+        """Actually make an HTTP request. This method exists only so the mock
+        can override it.
+        """
+        self._update_request_kwargs(kwargs)
+        return HTTP.request_with_timeout("POST", url, *args, **kwargs)
```
Comment (RishiDiwanTT) [inline api/millenium_patron.py:370]: Do we need to set a content-type as form encoded?

**Review by RishiDiwanTT — COMMENTED** [review]
One small comment, else all looks good :+1:

**Review by pcustic — COMMENTED** [review]
Mostly looks good 👍

File: api/millenium_patron.py (hunk 1 above)
Comment (io7m) [reply api/millenium_patron.py:370]: I'll check. Django didn't seem to need it, but I think it should be set.

File: api/millenium_patron.py [hunk 2]
```diff
@@ -57,6 +57,11 @@ class MilleniumPatronAPI(BasicAuthenticationProvider, XMLParser):
     # of the Millenium Patron API server.
     VERIFY_CERTIFICATE = "verify_certificate"
 
+    # Whether to use POST for remote requests. This is not supported by "standard"
+    # Millenium Patron API implementations; it is a LYRASIS extension to avoid having
```
[the remark this answers is not in the document]
Comment (io7m) [reply api/millenium_patron.py:61]: Yeah, might be right.

File: api/millenium_patron.py [hunk 3]
```diff
@@ -163,6 +168,18 @@ class MilleniumPatronAPI(BasicAuthenticationProvider, XMLParser):
...
             "default": BARCODE_FIELD,
         },
+        {
+            "key": USE_POST_FOR_REQUESTS,
+            "label": _("Use POST for requests"),
+            "description": _(
+                "Whether to use POST HTTP requests instead of GET. The Millenium Patron API has a design flaw in that "
+                "passwords appear in URLs made with GET requests. LYRASIS implementations of the API allow for using "
+                "HTTP POST so that passwords are no longer visible in URLs. If you are talking a LYRASIS implementation "
+                "such as VLC, you should enable this setting."
```
[the remark this answers is not in the document]
Comment (io7m) [reply api/millenium_patron.py:178]: The problem here is that allowing `POST` is effectively a proprietary extension that only we support and no other implementation of the Millenium Patron API does. I've just noticed a grammatical error too... `If you are talking a...`

[the remark this answers is not in the document]
Comment (io7m) [reply api/millenium_patron.py:178]: Er, the point being that it's inherently a technical option and one that you'd need to know that you could use (and why) before you turned it on.

**Review by pcustic — APPROVED** [review]
Seems fine, I'll approve but wait for the @dwilcox or @jonathangreen regarding Lyrasis vs TPP wording before merging.

**Pull request merged**